### ISSUE TITLE
Segment is not removed from SEGMENT_RUNNERS in the case when no JMX callback was received after segment repair was triggered

### ISSUE THREAD

**Issue by sergeyyegournov** [thelastpickle/cassandra-reaper#324]
Here is the log output for a given segment_id - what happens is that after segment repair times out the segment has NOT_STARTED state if there was no JMX callback from Cassandra which switches segment into RUNNING state - but the logic `SegmentRunner.processTriggeredSegment` does not seem to handle this correctly and does not remove segment from SEGMENT_RUNNERS. Thus subsequent attempts to repair this segment fail - first there is `i.c.s.SegmentRunner - SegmentRunner already exists for segment with ID: 50f4efc2-05e1-11e8-ba66-c58d985d11a3` - and then `Looks like segment has been running` which never stops

`
INFO   [2018-01-30 12:17:32,630] [cassampsimulator:50f4ef90-05e1-11e8-ba66-c58d985d11a3] i.c.s.RepairRunner - Next segment to run : 50f4efc2-05e1-11e8-ba66-c58d985d11a3
INFO   [2018-01-30 12:17:33,866] [cassampsimulator:50f4ef90-05e1-11e8-ba66-c58d985d11a3:50f4efc2-05e1-11e8-ba66-c58d985d11a3] i.c.s.SegmentRunner - It is ok to repair segment '50f4efc2-05e1-11e8-ba66-c58d985d11a3' on repair run with id '50f4ef90-05e1-11e8-ba66-c58d985d11a3'
INFO   [2018-01-30 12:17:33,889] [cassampsimulator:50f4ef90-05e1-11e8-ba66-c58d985d11a3:50f4efc2-05e1-11e8-ba66-c58d985d11a3] i.c.j.JmxProxy - Triggering repair of range (-2464893060303236334,-2380254007645758247] for keyspace "outbrain_document_features" on host 10.10.106.231, with repair parallelism parallel, in cluster with Cassandra version '3.11.1' (can use DATACENTER_AWARE 'true'), for column families: []
INFO   [2018-01-30 12:17:34,136] [cassampsimulator:50f4ef90-05e1-11e8-ba66-c58d985d11a3:50f4efc2-05e1-11e8-ba66-c58d985d11a3] i.c.s.SegmentRunner - Repair for segment 50f4efc2-05e1-11e8-ba66-c58d985d11a3 started, status wait will timeout in 1800000 millis
INFO   [2018-01-30 12:47:34,236] [cassampsimulator:50f4ef90-05e1-11e8-ba66-c58d985d11a3:50f4efc2-05e1-11e8-ba66-c58d985d11a3] i.c.s.SegmentRunner - Repair command 1038 on segment 50f4efc2-05e1-11e8-ba66-c58d985d11a3 returned with state NOT_STARTED
ERROR  [2018-01-30 12:47:34,313] [cassampsimulator:50f4ef90-05e1-11e8-ba66-c58d985d11a3:50f4efc2-05e1-11e8-ba66-c58d985d11a3] i.c.s.SegmentRunner - Segment 50f4efc2-05e1-11e8-ba66-c58d985d11a3 returned with startTime 2018-01-30T17:17:33.866Z and endTime null. This should not happen.Intensity cannot apply, so next run will start immediately.
INFO   [2018-01-30 13:18:46,213] [cassampsimulator:50f4ef90-05e1-11e8-ba66-c58d985d11a3] i.c.s.RepairRunner - Next segment to run : 50f4efc2-05e1-11e8-ba66-c58d985d11a3
ERROR  [2018-01-30 13:18:46,221] [cassampsimulator:50f4ef90-05e1-11e8-ba66-c58d985d11a3] i.c.s.SegmentRunner - SegmentRunner already exists for segment with ID: 50f4efc2-05e1-11e8-ba66-c58d985d11a3
io.cassandrareaper.ReaperException: SegmentRunner already exists for segment with ID: 50f4efc2-05e1-11e8-ba66-c58d985d11a3
INFO   [2018-01-30 13:18:46,222] [cassampsimulator:50f4ef90-05e1-11e8-ba66-c58d985d11a3] i.c.s.RepairRunner - Looks like segment #50f4efc2-05e1-11e8-ba66-c58d985d11a3 has been running more than an hour. Start time: 2018-01-30T17:17:33.866Z
INFO   [2018-01-30 13:18:46,224] [cassampsimulator:50f4ef90-05e1-11e8-ba66-c58d985d11a3] i.c.s.RepairRunner - Looks like segment #50f4efc2-05e1-11e8-ba66-c58d985d11a3 has been running more than an hour. Start time: 2018-01-30T17:17:33.866Z
INFO   [2018-01-30 13:18:46,225] [cassampsimulator:50f4ef90-05e1-11e8-ba66-c58d985d11a3] i.c.s.RepairRunner - Looks like segment #50f4efc2-05e1-11e8-ba66-c58d985d11a3 has been running more than an hour. Start time: 2018-01-30T17:17:33.866Z
INFO   [2018-01-30 13:18:46,227] [cassampsimulator:50f4ef90-05e1-11e8-ba66-c58d985d11a3] i.c.s.RepairRunner - Looks like segment #50f4efc2-05e1-11e8-ba66-c58d985d11a3 has been running more than an hour. Start time: 2018-01-30T17:17:33.866Z
INFO   [2018-01-30 13:18:46,228] [cassampsimulator:50f4ef90-05e1-11e8-ba66-c58d985d11a3] i.c.s.RepairRunner - Looks like segment #50f4efc2-05e1-11e8-ba66-c58d985d11a3 has been running more than an hour. Start time: 2018-01-30T17:17:33.866Z
`

Small update on this - since this issue happens consistently for one of the clusters I actually checked what Cassandra logs tell - the whole repair command seems to work successfully. However all repair sessions that suffer from the lack of JMX callback are repairing keyspaces that have just tiny amounts of data:

`
INFO  [Repair-Task-244] 2018-02-02 00:25:56,705 RepairRunnable.java:138 - Starting repair command #243 (8ba69d10-07d9-11e8-9aa1-d5789b179560), repairing keyspace outbrain_document with repair options (parallelism: parallel, primary range: false, incremental: false, job threads: 1, ColumnFamilies: [], dataCenters: [], hosts: [], # of ranges: 1, pull repair: false)
INFO  [InternalResponseStage:6840] 2018-02-02 00:25:56,779 RepairRunnable.java:342 - Repair command #243 finished in 0 seconds
`

This is a repair session which did not trigger JMX callback - I think it just completes too quick

**Comment by powellchristoph** [comment]
I am experiencing the same issue. It appears hung on that keyspace.

https://gist.github.com/powellchristoph/67fee503c3ea45e207e377de91e7334a

OS: Ubuntu 14.04 LTS
Java: version "1.8.0_151"
Reaper version: 1.0.2

cassandra-reaper.yaml
```
# Cassandra Reaper Configuration Example.
# See a bit more complete example in:
# src/server/src/test/resources/cassandra-reaper.yaml
segmentCount: 200
repairParallelism: PARALLEL
repairIntensity: 1.0
scheduleDaysBetween: 5
repairRunThreadCount: 15
hangingRepairTimeoutMins: 30
storageType: memory
enableCrossOrigin: true
incrementalRepair: false
enableDynamicSeedList: true
repairManagerSchedulingIntervalSeconds: 5
activateQueryLogger: false
jmxConnectionTimeoutInSeconds: 5
useAddressTranslator: false
datacenterAvailability: ALL

logging:
  level: INFO
  loggers:
    io.dropwizard: WARN
    org.eclipse.jetty: WARN
  appenders:
    - type: console
      logFormat: "%-6level [%d] [%t] %logger{5} - %msg %n"

server:
  type: default
  applicationConnectors:
    - type: http
      port: 8080
      bindHost: 0.0.0.0
  adminConnectors:
    - type: http
      port: 8081
      bindHost: 0.0.0.0
  requestLog:
    appenders: []

autoScheduling:
  enabled: true
  initialDelayPeriod: PT15S
  periodBetweenPolls: PT15S
  timeBeforeFirstSchedule: PT15S
  scheduleSpreadPeriod: PT1M
```

**Comment by sergeyyegournov** [comment]
@powellchristoph You're running 1.0.2 - and since then there were quite a few issues addressed which may cause repair sessions to get stuck. My case here is a bit special because it happens in a unlikely situation when JMX callback from Cassandra is not received. I suggest you build 1.1.0-SNAPSHOT from trunk or from PR branch and try again

**Comment by adejanovski** [comment]
@sergeyyegournov : it seems likely that you're getting JMX notifications that Reaper doesn't handle.
As you can see [here](https://github.com/thelastpickle/cassandra-reaper/blob/master/src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java#L714-L780), we fallback to debug logging when we get an unknown notification.

Could you try to reproduce the bug with debug logging enabled ?

**Comment by adejanovski** [comment]
Fixed by #329
